doc/architecture/decisions/0015-container-api.md

```diff
@@ -0,0 +1,103 @@
+# 15. Container API
+
+Date: 2018-09-17
+
+## Status
+
+Proposed
+
+## Context
+
+* We have several different roles for using a Holochain app (through a *container*):
+  * local, QML based UI components
+  * as a special case of the above: administration UI
+  * externally located (living in a browser) web UIs connected through some form of IPC
+  * *bridged* apps, i.e. Holochain app as the user of another Holochain app
+  * A Holo light-client connecting to the Holochain instance running on a HoloPort full-node
+* With HoloSqape and the Holochain-nodejs, we already have two container implementations
+
+A Holochain app cannot be used directly. Holochain is built as a library that exposes an interface for
+creating Holochain instances and calling their zome functions (see [core_api](/core_api/src/lib.rs)).
+This *core_api* requires its client to provide a context, which holds the representation of the agent
+(name and keys), a logger and persistence and will also include a networking proxy in the future.
+
+While it is possible to use this library directly in a monolithic app where the UI is tightly
+coupled to the app and everything linked to the Holochain library into the same executable, we regard this
+as a special case, since this does not allow for app composability (which we think is a crucial
+concept for the Holochain app ecosystem). 
```

Comment:
This sentence is not talking about UI composability, @ddd-mtl.
If you bundle the UI and Holochain System in an executable as in "use-case 1" (most left column in https://realtimeboard.com/app/board/o9J_kzZtCZI=/), you can't even have app composability because the app is hidden from the user and only usable through one UI. For app composability we need a (standardized) way to call zome functions from other apps (i.e. bridging) and a way for the user to set these up. This is what the Container is about. No reference to UI composability.

But maybe @zippy and @artbrock should give their approval for the importance of composability?

Reply:
The paragraph does mention UI and would not be necessary if you think of app as something without UI.
I guess there is a misunderstanding around two words: 'app', and 'composability'.
I understand app as final end-user facing app with UI, so app composability encompass UI composability.
I understand 'composability' as the capacity to let developers compose an app out of pre-existing holochain-apps (where holochain-app is just DNA, no UI). Scenario 1 does allows that.
I see how there can be a different definition of 'composability ' as the property of letting end-users compose their own app out of pre-existing 'app building blocks'. Scenario 1 indeed does not allow that.

Reply:
So for me this depends on the definition of "crucial"  I do think it's very important for long term power and viability, I'm not so sure it's next step on critical path for adoption.  @artbrock and I keep talking about "UI hints" as something to have inside the Genome that provide semantic intent for different types of UI.  That seems very important to land early(ish) because it signals the intent to composibility.  I agree that the clarity on the term app would help because if you look at it as the very end user experience at the end of all development, then yes it's not so much about composibility.  But if app refers to one of the microservices that may have a UI component also, well that would be pretty crucial to compose with, otherwise how are they microservices?

Reply:
Yeah, I was using the latter semantic for composability and I have (so far always) used "app" to mean "Holochain micro-service = instance of one Genome = no UI.

Reply:
So if I heard you right @zippy, you believe the important thing is to signal early that we intent holochain-apps to be UI-composable by developers making end-user facing apps?
@lucksus I understand you used 'app' for meaning 'holochain-app'. This is revealing to me that using terms like 'holochain-xxx' is prone to confusion, because "laziness" will push people to just say 'xxx' and ambiguity will start to form.